staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1091 @@
+	attGuestIP, attHostIP gonet.IP,
+	attVNI uint32,
+	attNSN k8stypes.NamespacedName) (existingIfc netfabric.NetworkInterface, attHasIfc bool, err error) {
+
+	if attState != nil {
+		existingIfc, attHasIfc = attState.ifc, attState.ifcIsValid
+	}
+
+	newIfcNeedsToBeCreated := !attHasIfc || ifcNeedsUpdate(existingIfc, attGuestIP, attHostIP, attVNI)
+
+	err = ca.deleteIfc(existingIfc, attHasIfc && newIfcNeedsToBeCreated)
+	if err != nil {
+		err = fmt.Errorf("Update of network interface of attachment %s failed, old interface %#+v could not be deleted: %s",
+			attNSN,
+			existingIfc,
+			err.Error())
+		return
+	}
+
+	if newIfcNeedsToBeCreated {
+		attHasIfc = false
+		guestMAC := generateMACAddr(attVNI, attGuestIP)
+		newIfc := netfabric.NetworkInterface{
+			Name:     generateIfcName(guestMAC),
+			VNI:      attVNI,
+			GuestMAC: guestMAC,
+			GuestIP:  attGuestIP,
+			HostIP:   attHostIP,
+		}
+		if attHostIP.Equal(ca.hostIP) {
+			err = ca.netFabric.CreateLocalIfc(newIfc)
+		} else {
+			err = ca.netFabric.CreateRemoteIfc(newIfc)
+		}
+		if err != nil {
+			err = fmt.Errorf("Creation of network interface of attachment %s failed, interface %#+v could not be created: %s",
+				attNSN,
+				newIfc,
+				err.Error())
+			return
+		}
+		existingIfc = newIfc
+		attHasIfc = true
+	}
+
+	return
+}
+
+func (ca *ConnectionAgent) deleteIfc(ifc netfabric.NetworkInterface, ifcNeedsDeletion bool) error {
+	var err error
+	if ifcNeedsDeletion {
+		if ifc.HostIP.Equal(ca.hostIP) {
+			// If we're here the interface is local
+			err = ca.netFabric.DeleteLocalIfc(ifc)
+		} else {
+			// If we're here the interface is remote
+			err = ca.netFabric.DeleteRemoteIfc(ifc)
+		}
+	}
+	return err
+}
+
+func (ca *ConnectionAgent) setHostIPAndIfcNameInLocalAttStatus(att *netv1a1.NetworkAttachment,
+	ifcName string,
+	ifcNameIsValid bool) (*netv1a1.NetworkAttachment, error) {
+
+	att2 := att.DeepCopy()
+	att2.Status.HostIP = ca.hostIP.String()
+	if ifcNameIsValid {
+		att2.Status.IfcName = ifcName
+	}
+	updatedAtt, err := ca.netv1a1Ifc.NetworkAttachments(att2.Namespace).Update(att2)
+	return updatedAtt, err
+}
+
+// removeAttFromVNState removes attName from the vnState associated with vni, both for local and remote
+// attachments. If attName is the last local attachment in the vnState, vnState is returned, so that
+// the caller can perform additional clean up (e.g. stopping the remote attachments informer).
+func (ca *ConnectionAgent) removeAttFromVNState(attName string, vni uint32) *vnState {
+	ca.vniToVnStateMutex.Lock()
+	defer ca.vniToVnStateMutex.Unlock()
+	vnState := ca.vniToVnState[vni]
+	if vnState != nil {
+		delete(vnState.localAtts, attName)
+		if len(vnState.localAtts) == 0 {
+			delete(ca.vniToVnState, vni)
+			return vnState
+		}
+		delete(vnState.remoteAtts, attName)
+	}
+	return nil
+}
+
+func (ca *ConnectionAgent) stopRemoteAttInformerAndEnqueueRemoteAttRefs(vnState *vnState, vni uint32) {
```

Comment:
This func's name is overly specific.  It would be better named something simpler, such as `closeVNState` or `finishVN`.

Reply:
Renamed to "clearVNResources" in d077e4f.